components/builder-scheduler/src/server/logger.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+use chrono::prelude::*;
+
+use protocol::jobsrv::Job;
+use protocol::scheduler::{Group, Project};
+
+pub struct Logger {
+    file: File,
+}
+
+impl Logger {
+    pub fn init<T: AsRef<Path>>(log_path: T) -> Self {
+        let filepath = log_path
+            .as_ref()
+            .to_path_buf()
+            .join("builder-scheduler.log");
+        Logger {
+            file: File::create(filepath).expect("Failed to initialize builder-scheduler log file"),
+        }
+    }
+
+    pub fn log(&mut self, msg: &str) {
+        let dt: DateTime<UTC> = UTC::now();
+        let fmt_msg = format!("{},{}\n", dt.format("%Y-%m-%d %H:%M:%S"), msg);
+
+        self.file
+            .write_all(fmt_msg.as_bytes())
+            .expect(&format!("Logger unable to write to {:?}", self.file));
+    }
+
+    // Log format (fields are comma-separated)
+    //   Log entry datetime (UTC)
+    //   Entry type - G (group), J (job), P (project)
+    //   Group id
+    //   State
+    //   Project name (for job or project)
+    //   Start datetime (UTC) (only for jobs)
+    //   End datetime (UTC) (only for jobs)
+    //   Start offset (offset from group creation, in seconds, only for jobs)
+    //   Duration (job duration, in seconds, only for jobs)
```

Comment:
Thinking a bit further out, I think it would be a good idea to think about using structured logging (e.g. JSON), sending everything to standard output, 12-factor app style, and using an ELK stack (or similar) to aggregate and analyze this kind of output.

Defining a type for each log event, with appropriate `From` trait implementations, would be an interesting way to implement things here, and could make a future transition to structured JSON logging pretty unintrusive. It would also very clearly and unambiguously document the contract of each log event.